031-statefulset-removal.md

```diff
@@ -0,0 +1,225 @@
+They will identify which pods need to be rolled and delete them.
+The only difference will be that instead of the Kubernetes StatefulSet controller, it will be the KafkaNodeSet controller which will create a new pod.
+For each Pod managed by a KafkaNodeSet, Strimzi will calculate a revision (really a _hash_ of the Pod configuration).
+This revision will be used to identify changes to the desired Pod configuration and trigger rolling of the pod.
+This is a similar process as used by the StatefulSet itself.
+
+Similarly to the StatefulSets, the KafkaNodeSets will be stable resources as well.
+All the annotations currently used on StatefulSets will be applied to them.
+This applies to the annotations used to store the configurations and state but also to the command annotations such as manual rolling update triggers.
+
+### Feature Gate
+
+The switch from StatefulSets to KafkaNodeSets should be controller by a new feature gate called `StatefulSetRemoval`.
+With this feature gate disabled, StatefulSets will be used.
+With it enabled, KafkaNodeSets will be used.
+The following table shows the expected graduation of the feature gate:
+
+| Phase | Strimzi versions       | Default state                                          |
+|:------|:-----------------------|:-------------------------------------------------------|
+| Alpha | 0.27, 0.28             | Disabled by default                                    |
+| Beta  | 0.29, 0.30             | Enabled by default                                     |
+| GA    | 0.31 and newer         | Enabled by default (without possibility to disable it) |
+
+As always, the feature gate timeline is subject to change based on the actual progress.
+If needed, it can be also removed.
+The timeline is relatively aggressive, but feature gating all the development will not be trivial.
+So we should try to move quickly.
+
+### Implementation
+
+In the first phase, we should implement the 1-to-1 replacement for StatefulSets.
+While that will not bring any immediate advantages, it will allow us to easily move from StatefulSets to KafkaNodeSets.
+It would also avoid big PRs and instead allow us to gradually improve in smaller steps.
+
+### Risks
+
+In the current solution, Kubernetes is responsible for (re)starting the Pods in case of disruptions.
+After implementing this proposal, Strimzi will be responsible for it.
+Any bugs and issues in Strimzi might result in the Pods not being started and might affect the availability of the Kafka clusters.
+By accepting this proposal, we need to be able to accept this risk and responsibility.
+
+Unlike StatefulSets, KafkaNodeSet also doesn't use template to generate Pods.
+That brings us many advantages such as asymmetric configuration etc.
+But it also means that the custom resource will be bigger than StatefulSets because it contains a full YAML for each Pod.
+It will have a limited capacity how many Pods it will be able to handle.
```

Comment:
Can we estimate, even if very roughly, what this limit is likely to be? Are we talking about 10s of pods, or 100s?

Reply:
No, I do not have any estimate.

Reply:
Well the bound is on the size of CRs supported by kubernetes (which IIUC is inherited from the etcd limit of 1.5MB by default on key-value pairs). So presumably we could estimate this by taking a Kafka `Pod` spec (for a cluster using all the bells and whistles) and seeing how many bytes it requires.

Reply:
Oh, I first thought you are asking about the performance and not about how many pods can fit. 

IIRC the limit is 1MB. But it depends heavily how you define _cluster using all the bells and whistles_. How much metadata does that mean? How many affinity rules? How many volumes does it have? How many topology spread-rules? My ZooKeeper pods have about 4KB each. It has some metadata, but no affinity, spread constraints etc.

Kafka pods will be probably bigger with more volumes etc. ... so maybe 100 pods?